File: project_manager_vue3/src/api/todo.ts
import request from './request'

export interface TodoItem {
  id: number
  project_id: number
  project_title: string
  description: string
  step_ids: number[]
  step_names: string[]
  completion_note?: string
  is_completed: boolean
  target_date?: string
  created_at: string
  updated_at: string
  student_name?: string  // 添加学生姓名字段
}

export interface TodoCalendarItem {
  date: string
  todo_count: number
  completed_count: number
  revenue?: number  // 每日收入
}

export interface TodoCreate {
  project_id: number
  description: string
  step_ids: number[]
  target_date?: string
}

export interface TodoUpdate {
  description?: string
  completion_note?: string
  is_completed?: boolean
  target_date?: string
  attachment_ids?: number[]  // 附件ID列表
}

export const todoApi = {
  // 获取指定日期的待办列表
  getTodos: (targetDate?: string) => {
    const params: any = {}
    if (targetDate) {
      params.target_date = targetDate
    }
    return request.get<TodoItem[]>('/todos/', { params })
  },
  
  // 创建待办
  createTodo: (data: TodoCreate) => {
    return request.post<TodoItem>('/todos/', data)
  },
  
  // 更新待办
  updateTodo: (todoId: number, data: TodoUpdate) => {
    return request.put<TodoItem>(`/todos/${todoId}`, data)
  },
  
  // 删除待办
  deleteTodo: (todoId: number) => {
    return request.delete(`/todos/${todoId}`)
  },
  
  // 获取待办日历数据
  getCalendar: (year?: number, month?: number) => {
    const params: any = {}
    if (year) params.year = year
    if (month) params.month = month
    return request.get<TodoCalendarItem[]>('/todos/calendar', { params })
  },
}

File: fastapi_back/app/models/todo.py
"""
待办数据模型（ORM + DTO）

包含数据库表定义和DTO。
"""
from sqlmodel import SQLModel, Field, Relationship
from typing import Optional, List, TYPE_CHECKING
from datetime import datetime, date

if TYPE_CHECKING:
    from app.models.project import Project
    from app.models.historical_project import HistoricalProject


class Todo(SQLModel, table=True):
    """待办表"""
    __tablename__ = "todo"

    id: Optional[int] = Field(default=None, primary_key=True)
    project_id: Optional[int] = Field(default=None, foreign_key="project.id", description="所属项目ID")
    historical_project_id: Optional[int] = Field(default=None, foreign_key="historicalproject.id", description="所属历史项目ID")
    description: str = Field(description="待办描述")
    step_ids: str = Field(description="步骤ID列表，JSON格式字符串")
    completion_note: Optional[str] = Field(default=None, description="完成说明")
    is_completed: bool = Field(default=False, description="是否已完成")
    target_date: Optional[datetime] = Field(default=None, description="目标日期")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow, sa_column_kwargs={"onupdate": datetime.utcnow})

    # 关系
    project: Optional["Project"] = Relationship(back_populates="todos")
    historical_project: Optional["HistoricalProject"] = Relationship(back_populates="todos")


# DTO类（保持向后兼容）
class TodoBase(SQLModel):
    """待办基础模型"""
    description: str
    step_ids: List[int] = []
    completion_note: Optional[str] = None
    is_completed: bool = False
    target_date: Optional[datetime] = None


class TodoCreate(TodoBase):
    """创建待办"""
    project_id: Optional[int] = None
    historical_project_id: Optional[int] = None


class TodoUpdate(SQLModel):
    """更新待办"""
    description: Optional[str] = None
    step_ids: Optional[List[int]] = None
    completion_note: Optional[str] = None
    is_completed: Optional[bool] = None
    target_date: Optional[datetime] = None
    attachment_ids: Optional[List[int]] = None


class TodoRead(TodoBase):
    """读取待办"""
    id: int
    project_id: Optional[int] = None
    historical_project_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TodoReadWithRelations(TodoRead):
    """包含关联的待办读取"""
    project_title: Optional[str] = None
    step_names: List[str] = []
    student_name: Optional[str] = None

